ADO.NET/OAuthSampleApp/OAuthSampleApp.cs

```diff
@@ -0,0 +1,216 @@
+    private const string OAUTH_REFRESH_TOKEN = "OAUTH_SAMPLE_REFRESH_TOKEN";
+
+    static async Task Main(string[] args)
+    {
+        connectionStringBuilder = new VerticaConnectionStringBuilder();
+        connectionStringBuilder.ConnectionString = ConfigurationManager.ConnectionStrings["ConnectionString"].ConnectionString;
+
+        await SetUp();
+
+        // change the user from the dbadmin to OAuth user
+        connectionStringBuilder.User = ConfigurationManager.AppSettings["User"];
+
+        await EnsureAccessToken();
+        await ConnectToDatabase();
+        await TearDown();
+    }
+
+    // admin must log in and setup the user for OAuth
+    // steps can also be followed in vsql
+    static async Task SetUp()
+    {
+        vConnection = new VerticaConnection(connectionStringBuilder.ConnectionString);
+        vConnection.Open();
+
+        string USER = ConfigurationManager.AppSettings["User"];
+        string CLIENT_ID = ConfigurationManager.AppSettings["ClientId"];
+        string CLIENT_SECRET = ConfigurationManager.AppSettings["ClientSecret"];
+        string INTROSPECT_URL = ConfigurationManager.AppSettings["TokenUrl"] + "introspect";
+
+        VerticaCommand cmd = new VerticaCommand("DROP USER IF EXISTS " + USER + " CASCADE", vConnection);
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "DROP AUTHENTICATION IF EXISTS adooauth CASCADE";
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "CREATE USER " + USER;
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "CREATE AUTHENTICATION adooauth METHOD 'oauth'" + GetAuthHost();
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "GRANT AUTHENTICATION adooauth TO " + USER;
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "GRANT ALL ON SCHEMA PUBLIC TO " + USER;
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "ALTER AUTHENTICATION adooauth SET client_id = '" + CLIENT_ID + "'";
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "ALTER AUTHENTICATION adooauth SET client_secret = '" + CLIENT_SECRET + "'";
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "ALTER AUTHENTICATION adooauth SET introspect_url = '" + INTROSPECT_URL + "'";
+        cmd.ExecuteNonQuery();
+    }
+
+    // removes the oauth user and closes the connection for the admin
+    public static async Task TearDown()
+    {
+        string USER = ConfigurationManager.AppSettings["User"];
+        VerticaCommand cmd = new VerticaCommand("DROP USER IF EXISTS " + USER + " CASCADE", vConnection);
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "DROP AUTHENTICATION IF EXISTS adooauth CASCADE";
+        cmd.ExecuteNonQuery();
+
+        vConnection.Close();
+    }
+
+    // if no access token is set, obtains and sets first access and refresh tokens
+    static async Task EnsureAccessToken()
+    {
+        string accessToken = Environment.GetEnvironmentVariable(OAUTH_ACCESS_TOKEN, EnvironmentVariableTarget.User);

```

Comment:
Suggestion:  You could make the accessToken and refreshToken static member variables and set them like this at the beginning of the Main method.  At the end of the Main method it could set the variables for the next run of the program.   I think this would clarify why we are using the environment variables.

Reply:
I think this would make sense in a test, but not in a sample application. If I was looking at this as a first time OAuth user, I would ask "why are trying to get a token when it doesn't exist?" and "why are we setting a token at the end of the program when we don't need it anymore?"